llvm/tools/plugins-shlib/pypass.cpp

```diff
@@ -0,0 +1,517 @@
+    if (!resolveSymbols())
+      return;
+    Py_InitializeEx(0);
+    PyBuiltins = PyImport_ImportModule("builtins");
+    PyGlobals = PyModule_GetDict(PyImport_AddModule("__main__"));
+    Valid = true;
+  }
+
+  ~PythonAPI() {
+    if (std::atomic_exchange(&Valid, false)) {
+      Py_DecRef(PyBuiltins);
+      Py_DecRef(PyGlobals);
+      Py_FinalizeEx();
+    }
+  }
+
+  bool loadDylib(std::string Path) {
+    // Empty paths dlopen host process
+    if (Path.empty()) {
+      errs() << "Failed to load Python shared library: ''\n";
+      return false;
+    }
+    std::string Err;
+    Dylib = sys::DynamicLibrary::getPermanentLibrary(Path.c_str(), &Err);
+    if (!Dylib.isValid()) {
+      errs() << "Failed to load Python shared library: '" << Path << "' ("
+             << Err << ")\n";
+      return false;
+    }
+
+    return true;
+  }
+
+  bool resolveSymbols() {
+    bool Success = true;
+    Success &= resolve("_Py_IncRef", &Py_IncRef);
+    Success &= resolve("_Py_DecRef", &Py_DecRef);
+    Success &= resolve("Py_InitializeEx", &Py_InitializeEx);
+    Success &= resolve("Py_FinalizeEx", &Py_FinalizeEx);
+    Success &= resolve("PyErr_Print", &PyErr_Print);
+    Success &= resolve("PyGILState_Ensure", &PyGILState_Ensure);
+    Success &= resolve("PyGILState_Release", &PyGILState_Release);
+    Success &= resolve("PyImport_AddModule", &PyImport_AddModule);
+    Success &= resolve("PyImport_ImportModule", &PyImport_ImportModule);
+    Success &= resolve("PyModule_GetDict", &PyModule_GetDict);
+    Success &= resolve("PyDict_GetItemString", &PyDict_GetItemString);
+    Success &= resolve("PyDict_SetItemString", &PyDict_SetItemString);
+    Success &= resolve("PyDict_New", &PyDict_New);
+    Success &= resolve("PyObject_CallObject", &PyObject_CallObject);
+    Success &= resolve("PyObject_GetAttrString", &PyObject_GetAttrString);
+    Success &= resolve("PyObject_IsTrue", &PyObject_IsTrue);
+    Success &= resolve("PyLong_FromVoidPtr", &PyLong_FromVoidPtr);
+    Success &= resolve("PyUnicode_FromString", &PyUnicode_FromString);
+    Success &= resolve("PyTuple_SetItem", &PyTuple_SetItem);
+    Success &= resolve("PyTuple_New", &PyTuple_New);
+    return Success;
+  }
+
+  bool importModule(const char *Name) const {
+    void *Mod = PyImport_ImportModule(Name);
+    if (!Mod) {
+      PyErr_Print();
+      return false;
+    }
+    PyDict_SetItemString(PyGlobals, Name, Mod);
+    Py_DecRef(Mod);
+    return true;
+  }
+
+  bool evaluate(std::string Code, void *Globals) const {
+    void *Exec = PyObject_GetAttrString(PyBuiltins, "exec");
+    void *Args = PyTuple_New(2);
+    if (!Args)
+      return false;
+    if (PyTuple_SetItem(Args, 0, PyUnicode_FromString(Code.c_str())))
+      return false;
+    Py_IncRef(Globals);
+    if (PyTuple_SetItem(Args, 1, Globals))
+      return false;
+
+    // Interpreter is not thread-safe
+    auto GIL = make_scope_exit(
+        [this, Lock = PyGILState_Ensure()]() { PyGILState_Release(Lock); });
```

Comment:
Shouldn’t the GIL be acquired prior to creating Python objects, at the very beginning?